Add unbound integration

dbyron0 opened the pull request:
### What does this PR do?

This adds support for unbound DNS caching resolver. http://unbound.net/

Originally written by @joewilliams (DataDog/dd-agent#2853)

and then continued in https://github.com/DataDog/integrations-extras/pull/41, and now here we are.

### Motivation

We need the metrics.

dbyron0 commented:
OK, looks like travis failures are related to the code here.  I'll dig in and see what I can figure out.

joewilliams commented:
Thanks for taking over this plugin! I no longer focus on DNS at work so I haven't had time work on it.

dabcoder commented:
@dbyron0 Thanks for continuing to work on this. The tests are currently failing because of `datadog_checks.base.errors.ConfigurationError: 'unbound-control' executable not found` (it's not installed in the Travis [Docker] container which is based on the `Xenial` Ubuntu image). You'll probably want to mock the output of the command instead. 

I can potentially help with that when the check implementation is done, let me know.

dbyron0 commented:
I was going to try setting up an integration test w/ docker compose to see how well the check is actually working.  Would love help with a mock in the meantime though.

dbyron0 commented:
I think I broke the rules by using `f''` strings.  Do I need this code to work with both 2.7 and 3.x ?

dbyron0 commented:
Looks like https://docs.datadoghq.com/developers/integrations/python/#python-2-or-python-3 answers my question... yes...compatible with both.

apigirl commented on unbound/README.md:
```diff
@@ -0,0 +1,51 @@
+
+### Installation
+
+The Unbound check is not included in the [Datadog Agent][2] package, so it must
```
Can you please add more specific instructions about how you would install it? The [portworx integration]9https://github.com/DataDog/integrations-extras/tree/master/portworx) is a great example of this

dbyron0 replied:
Sure thing.  I have a bit more work left on the code, but this is certainly on my list.  Thanks for the pointer to a good example.

dbyron0 replied:
How's this? https://github.com/DataDog/integrations-extras/pull/451/commits/888b116deca8d23458e17f5ca50696f5d98cfd1a

l0k0ms commented on unbound/datadog_checks/unbound/data/conf.yaml.example:
```diff
@@ -0,0 +1,44 @@
+    #
+    # config_file: /path/to/unbound.conf
+
+    ##
```
```suggestion

```

dbyron0 commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,41 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
+unbound.thread0.num.queries,rate,,,,number of queries received by thread,1,unbound,thread0.num.queries
```
If this is supposed to list every possible metric this could generate, I'm in a bit of a pickle....or maybe I need to structure the code differently.  The number of threads is configurable, and could be big, so the way the code is currently written we could have unbound.thread0.*, unbound.thread1.*, unbound.thread2.*, etc.  I suppose this means I need to pick off the thread number and use it as a tag...How do other folks handle this?

dbyron0 replied:
Gave it a shot in https://github.com/DataDog/integrations-extras/pull/451/commits/ece0b148f971b032da36c4c3b11d516a314b2c20

dabcoder replied:
> I suppose this means I need to pick off the thread number and use it as a tag.

Yes, using tags for this case seems like a good idea.

dbyron0 replied:
Found a few more cases like this that I still need to address.

dbyron0 replied:
Done with this.

dabcoder commented:
@dbyron0 thanks for following up. 

The tests are now failing because a `service_checks.json` file is missing. Would this integration need to send any (service checks)? If so you can refer to: https://docs.datadoghq.com/developers/integrations/new_check_howto/#service-check-file. If not, you'll still need to add the file which can contain an empty list `[]`.

dbyron0 commented:
@dabcoder I'm torn about the service check.  Seems like having one would add some value, but I'm trying to figure out ow to do it safely.  I haven't found a great explanation of what a service check is, or the potential status codes, and what happens if I report something other than OK.  From grepping through the code I found OK, WARNING, CRITICAL and UNKNOWN as the options.

As for implementing one, I'm pondering these options:
* 0 exitcode from the unbound-control stats command means ok, non-zero means critical? or maybe warning?
* exception from unbound-control stats command means unknown, or warning?
* run a separate unbound-control status command and use its exit code/exception.  Leaning away from this though.

But, unbound-control could fail because of some config struggle or other problem where unbound itself is working fine....so maybe that argues for only using warning if unbound-control fails.

Or maybe it's better to depend on some lower level thing like systemd and assume there are metrics/service checks there?  Or leave it to a dns client to do periodic DNS lookups as a health check?  And those choices mean no service check here...

dabcoder commented:
@dbyron0 

> As for implementing one, I'm pondering these options:
> 0 exitcode from the unbound-control stats command means ok, non-zero means critical? or maybe warning?

I would say non-zero would mean critical if the stats command is unable to provide any metrics at all.

> exception from unbound-control stats command means unknown, or warning?

This could be a warning indeed. 

> run a separate unbound-control status command and use its exit code/exception. Leaning away from this though

I would also say that this is not necessary here.

> But, unbound-control could fail because of some config struggle or other problem where unbound itself is working fine....so maybe that argues for only using warning if unbound-control fails.

In case of configuration issues, you can log a message at info level I suppose, with `self.log.info(...)`, that would help users to fix the config, without the need to send any service checks there. 

I hope that helps?

dbyron0 commented:
It's hard to distinguish a config issue from a 'real' failure though...but it all comes down to documentation, right?  So that people seeing the issue know how to interpret it?

And, if the inability to get any stats means critical, then I'd say a non-zero exit code, an exception, or a zero exit code but no metrics all mean critical.

dbyron0 commented:
@dabcoder I could use a hand working out the failures...I've added an empty service_checks.json (at least for now) and
```
$ ddev validate manifest -i
```
passes for me locally.

dabcoder commented:
> @dabcoder I could use a hand working out the failures...I've added an empty `service_checks.json` (at least for now) 

Sorry, just seeing your message now. So the structure has changed fairly recently, here is what needs to be done:

- There needs to be an `assets` folder in the integration main directory. Example here: https://github.com/DataDog/integrations-extras/tree/master/hbase_master/assets. 
- The `service_checks.json` file need to be in that above `assets` directory.
- It then needs to be referenced in the `manifests.json` file: https://github.com/DataDog/integrations-extras/blob/d3aad6dde5a554f106fc39b66aba080464b66aba/hbase_master/manifest.json#L23-L27. You can leave monitors and dashboards as empty dicts like in this hbase integration.

And finally rebase on master. Let us know if the tests are still failing, which they shouldn't. Thanks again for following up.

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,100 @@
+unbound.msg.cache.count,gauge,,,,The number of items (DNS replies) in the message cache,0,unbound,msg.cache.count
+unbound.rrset.cache.count,gauge,,,,The number of RRsets in the rrset cache,0,unbound,rrset.cache.count
+unbound.infra.cache.count,gauge,,,,The number of items in the infra cache,0,unbound,infra.cache.count
+unbound.key.cache.count,gauge,,,,The number of items in the key cache,0,key.cache.count
```
```suggestion
unbound.key.cache.count,gauge,,,,The number of items in the key cache,0,unbound,key.cache.count
```

dbyron0 commented:
I'm off for a couple of weeks though, but will get to this soon after I'm back.

rebase is also going to involve force pushing...seems like before there was some desire to avoid that.  Not sure how much of the history you really want to keep here.  I'm OK with nuking it all, but want to give credit to the other folks that have worked on this.

dabcoder commented:
> rebase is also going to involve force pushing...seems like before there was some desire to avoid that. Not sure how much of the history you really want to keep here

Rebase on `origin/master` should keep the history, as far as I know. It should add your work (and previous work) on top of the previous commits on master, so I think you should be fine there.

dbyron0 commented:
OK, I think I have the mechanical stuff taken care of.  Still some more code changes to make but I'm on the road...Thanks much.

dabcoder commented on unbound/datadog_checks/unbound/unbound.py:
```diff
@@ -0,0 +1,135 @@
+
+        """
+        Example of unbound stats outpout:
+        total.num.queries=12
```
nit: I had originally written this comment to get started with updating the previous out of date branch, maybe we could either remove it or move it somewhere closer to the stats output rather than at the beginning of the check?

dbyron0 replied:
Done in https://github.com/DataDog/integrations-extras/pull/451/commits/64d52b26358af85418cee939dac4b62dde98add5

ruthnaebeck commented on unbound/README.md:
```diff
@@ -0,0 +1,62 @@
+
+### Service Checks
+
+This check does not include any service checks.
```
```suggestion
The Unbound check does not include any service checks.
```

dbyron0 replied:
Took a crack at adding a service check in https://github.com/DataDog/integrations-extras/pull/451/commits/2af5e60f9b83ad20fc9451126345d71c27a18f35

ruthnaebeck commented on unbound/README.md:
```diff
@@ -0,0 +1,62 @@
+
+### Events
+
+This check does not include any events.
```
```suggestion
The Unbound check does not include any events.
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,100 @@
+metric_name,metric_type,interval,unit_name,per_unit_name,description,orientation,integration,short_name
```
Any metrics described as `number of queries` should be updated to:
metrics_type - `count`
interval - Add the interval if known
unit_name - `query`
description - `The number of queries...`

Any metrics described as `number of <something>` should be updated to:
metrics_type - `count`
interval - Add the interval if known
description - `The number of <something>...`

dbyron0 replied:
Took a crack at this in https://github.com/DataDog/integrations-extras/pull/451/commits/187abc7aa1d94e8e4acd0340799480156ba34203.

dabcoder commented:
> Still some more code changes to make

Great, let us know when this is ready for review code wise then. Thanks!

dbyron0 commented:
Getting closer.  Still going to try to add an integration test with an unbound docker image.  Fingers crossed that doesn't change the collector that much.

ruthnaebeck commented on unbound/README.md:
```diff
@@ -0,0 +1,62 @@
+
+### Service Checks
+
+**unbound.can_get_stats** : Returns CRITICAL if unbound-control fails or there's an error parsing its output.  Returns OK otherwise.
```
```suggestion
**unbound.can_get_stats**  
Returns CRITICAL if unbound-control fails or there's an error parsing its output.  Returns OK otherwise.
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.num.dnscrypt.crypted,count,,query,,The number of queries that were encrypted and successfully decapsulated by dnscrypt,1,unbound,thread.num.dnscrypt.cert
+unbound.thread.num.dnscrypt.cert,count,,query,,The number of queries that were requesting dnscrypt certificates,1,unbound,thread.num.dnscrypt.cert
+unbound.thread.num.dnscrypt.cleartext,count,,query,,The number of queries received on dnscrypt port that were cleartext and not a request for certificates,1,unbound,thread.num.dnscrypt.cleartext
+unbound.thread.num.dnscrypt.malformed,count,,,,The number of requests that were neither cleartext nor valid dnscrypt messages,1,unbound,thread.num.dnscrypt.malformed
```
```suggestion
unbound.thread.num.dnscrypt.malformed,count,,request,,The number of requests that were neither cleartext nor valid dnscrypt messages,1,unbound,thread.num.dnscrypt.malformed
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.num.prefetch,count,,,,The number of cache prefetches performed,1,unbound,thread.num.prefetch
+unbound.thread.num.zero_ttl,count,,,,The number of replies with ttl zero because they served an expired cache entry,1,unbound,thread.num.zero_ttl
+unbound.thread.num.recursivereplies,count,,,,The number of replies sent to queries that needed recursive processing,1,unbound,thread.num.recursivereplies
+unbound.thread.requestlist.avg,gauge,,,,The average number of requests in the  internal  recursive  processing request list on insert of a new incoming recursive processing query,0,unbound,thread.requestlist.avg
```
```suggestion
unbound.thread.requestlist.avg,gauge,,request,,The average number of requests in the internal recursive processing request list on insert of a new incoming recursive processing query,0,unbound,thread.requestlist.avg
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.num.zero_ttl,count,,,,The number of replies with ttl zero because they served an expired cache entry,1,unbound,thread.num.zero_ttl
+unbound.thread.num.recursivereplies,count,,,,The number of replies sent to queries that needed recursive processing,1,unbound,thread.num.recursivereplies
+unbound.thread.requestlist.avg,gauge,,,,The average number of requests in the  internal  recursive  processing request list on insert of a new incoming recursive processing query,0,unbound,thread.requestlist.avg
+unbound.thread.requestlist.max,gauge,,,,Maximum size attained by the internal recursive processing request list,1,unbound,thread.requestlist.max
```
```suggestion
unbound.thread.requestlist.max,gauge,,request,,Maximum size attained by the internal recursive processing request list,1,unbound,thread.requestlist.max
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.num.recursivereplies,count,,,,The number of replies sent to queries that needed recursive processing,1,unbound,thread.num.recursivereplies
+unbound.thread.requestlist.avg,gauge,,,,The average number of requests in the  internal  recursive  processing request list on insert of a new incoming recursive processing query,0,unbound,thread.requestlist.avg
+unbound.thread.requestlist.max,gauge,,,,Maximum size attained by the internal recursive processing request list,1,unbound,thread.requestlist.max
+unbound.thread.requestlist.overwritten,count,,,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
```
```suggestion
unbound.thread.requestlist.overwritten,count,,request,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.requestlist.avg,gauge,,,,The average number of requests in the  internal  recursive  processing request list on insert of a new incoming recursive processing query,0,unbound,thread.requestlist.avg
+unbound.thread.requestlist.max,gauge,,,,Maximum size attained by the internal recursive processing request list,1,unbound,thread.requestlist.max
+unbound.thread.requestlist.overwritten,count,,,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
+unbound.thread.requestlist.exceeded,gauge,,,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
```
```suggestion
unbound.thread.requestlist.exceeded,gauge,,query,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.requestlist.max,gauge,,,,Maximum size attained by the internal recursive processing request list,1,unbound,thread.requestlist.max
+unbound.thread.requestlist.overwritten,count,,,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
+unbound.thread.requestlist.exceeded,gauge,,,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
+unbound.thread.requestlist.current.all,gauge,,,,Current size of the request list,0,unbound,thread.requestlist.current
```
Size as in number of requests or size as in bytes?

dbyron0 replied:
number of requests.

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.requestlist.overwritten,count,,,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
+unbound.thread.requestlist.exceeded,gauge,,,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
+unbound.thread.requestlist.current.all,gauge,,,,Current size of the request list,0,unbound,thread.requestlist.current
+unbound.thread.requestlist.current.user,gauge,,,,Current size of the request list (only the requests from client queries),0,unbound,thread.requestlist.current.user
```
Size as in number of requests or size as in bytes?

dbyron0 replied:
number of requests.

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.requestlist.exceeded,gauge,,,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
+unbound.thread.requestlist.current.all,gauge,,,,Current size of the request list,0,unbound,thread.requestlist.current
+unbound.thread.requestlist.current.user,gauge,,,,Current size of the request list (only the requests from client queries),0,unbound,thread.requestlist.current.user
+unbound.thread.recursion.time.avg,gauge,,,,Average time it took to answer queries that needed recursive processing,0,unbound,thread.recursion.time.avg
```
Is this measured in milliseconds, seconds, other?

dbyron0 replied:
From https://github.com/NLnetLabs/unbound/blob/master/daemon/remote.c#L756, seconds

dbyron0 replied:
https://github.com/DataDog/integrations-extras/pull/451/commits/4363b6f34c3e4e164b0dd53d07fd518388100597

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.recursion.time.avg,gauge,,,,Average time it took to answer queries that needed recursive processing,0,unbound,thread.recursion.time.avg
+unbound.thread.recursion.time.median,gauge,,,,The median of the time it took to answer queries that needed recursive processing,0,unbound,thread.recursion.time.median
+unbound.thread.tcpusage,gauge,,,,The currently held tcp buffers for incoming connections,0,unbound,thread.tcpusage
+unbound.total.num.queries,count,,,,summed over threads,1,unbound,total.num.queries
```
```suggestion
unbound.total.num.queries,count,,query,,The total number of queries summed over threads,1,unbound,total.num.queries
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.recursion.time.median,gauge,,,,The median of the time it took to answer queries that needed recursive processing,0,unbound,thread.recursion.time.median
+unbound.thread.tcpusage,gauge,,,,The currently held tcp buffers for incoming connections,0,unbound,thread.tcpusage
+unbound.total.num.queries,count,,,,summed over threads,1,unbound,total.num.queries
+unbound.total.num.cachehits,count,,,,summed over threads,1,unbound,total.num.cachehits
```
```suggestion
unbound.total.num.cachehits,count,,hit,,The total number of cache hits summed over threads,1,unbound,total.num.cachehits
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.tcpusage,gauge,,,,The currently held tcp buffers for incoming connections,0,unbound,thread.tcpusage
+unbound.total.num.queries,count,,,,summed over threads,1,unbound,total.num.queries
+unbound.total.num.cachehits,count,,,,summed over threads,1,unbound,total.num.cachehits
+unbound.total.num.cachemiss,count,,,,summed over threads,1,unbound,total.num.cachemiss
```
```suggestion
unbound.total.num.cachemiss,count,,miss,,The total number of cache misses summed over threads,1,unbound,total.num.cachemiss
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.queries,count,,,,summed over threads,1,unbound,total.num.queries
+unbound.total.num.cachehits,count,,,,summed over threads,1,unbound,total.num.cachehits
+unbound.total.num.cachemiss,count,,,,summed over threads,1,unbound,total.num.cachemiss
+unbound.total.num.dnscrypt.crypted,count,,,,summed over threads,1,unbound,total.num.dnscrypt.crypted
```
```suggestion
unbound.total.num.dnscrypt.crypted,count,,,,The total number of dnscrypt crypted summed over threads,1,unbound,total.num.dnscrypt.crypted
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.cachehits,count,,,,summed over threads,1,unbound,total.num.cachehits
+unbound.total.num.cachemiss,count,,,,summed over threads,1,unbound,total.num.cachemiss
+unbound.total.num.dnscrypt.crypted,count,,,,summed over threads,1,unbound,total.num.dnscrypt.crypted
+unbound.total.num.dnscrypt.cert,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cert
```
```suggestion
unbound.total.num.dnscrypt.cert,count,,,,The total number of dnscrypt cert summed over threads,1,unbound,total.num.dnscrypt.cert
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.cachemiss,count,,,,summed over threads,1,unbound,total.num.cachemiss
+unbound.total.num.dnscrypt.crypted,count,,,,summed over threads,1,unbound,total.num.dnscrypt.crypted
+unbound.total.num.dnscrypt.cert,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cert
+unbound.total.num.dnscrypt.cleartext,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cleartext
```
```suggestion
unbound.total.num.dnscrypt.cleartext,count,,,,The total number of dnscrypt cleartext summed over threads,1,unbound,total.num.dnscrypt.cleartext
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.dnscrypt.crypted,count,,,,summed over threads,1,unbound,total.num.dnscrypt.crypted
+unbound.total.num.dnscrypt.cert,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cert
+unbound.total.num.dnscrypt.cleartext,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cleartext
+unbound.total.num.dnscrypt.malformed,count,,,,summed over threads,1,unbound,total.num.dnscrypt.malformed
```
```suggestion
unbound.total.num.dnscrypt.malformed,count,,,,The total number of dnscrypt malformed summed over threads,1,unbound,total.num.dnscrypt.malformed
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.dnscrypt.cert,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cert
+unbound.total.num.dnscrypt.cleartext,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cleartext
+unbound.total.num.dnscrypt.malformed,count,,,,summed over threads,1,unbound,total.num.dnscrypt.malformed
+unbound.total.num.prefetch,count,,,,summed over threads,1,unbound,total.num.prefetch
```
```suggestion
unbound.total.num.prefetch,count,,,,The total number of prefetch summed over threads,1,unbound,total.num.prefetch
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.dnscrypt.cleartext,count,,,,summed over threads,1,unbound,total.num.dnscrypt.cleartext
+unbound.total.num.dnscrypt.malformed,count,,,,summed over threads,1,unbound,total.num.dnscrypt.malformed
+unbound.total.num.prefetch,count,,,,summed over threads,1,unbound,total.num.prefetch
+unbound.total.num.zero_ttl,count,,,,summed over threads,1,unbound,total.num.zero_ttl
```
```suggestion
unbound.total.num.zero_ttl,count,,,,The total number of zero ttl summed over threads,1,unbound,total.num.zero_ttl
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.dnscrypt.malformed,count,,,,summed over threads,1,unbound,total.num.dnscrypt.malformed
+unbound.total.num.prefetch,count,,,,summed over threads,1,unbound,total.num.prefetch
+unbound.total.num.zero_ttl,count,,,,summed over threads,1,unbound,total.num.zero_ttl
+unbound.total.num.recursivereplies,count,,,,summed over threads,1,unbound,total.num.recursivereplies
```
```suggestion
unbound.total.num.recursivereplies,count,,,,The total number of recursive replies summed over threads,1,unbound,total.num.recursivereplies
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.prefetch,count,,,,summed over threads,1,unbound,total.num.prefetch
+unbound.total.num.zero_ttl,count,,,,summed over threads,1,unbound,total.num.zero_ttl
+unbound.total.num.recursivereplies,count,,,,summed over threads,1,unbound,total.num.recursivereplies
+unbound.total.requestlist.avg,gauge,,,,averaged over threads,0,unbound,total.requestlist.avg
```
```suggestion
unbound.total.requestlist.avg,gauge,,request,,The total number of requests averaged over threads,0,unbound,total.requestlist.avg
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.zero_ttl,count,,,,summed over threads,1,unbound,total.num.zero_ttl
+unbound.total.num.recursivereplies,count,,,,summed over threads,1,unbound,total.num.recursivereplies
+unbound.total.requestlist.avg,gauge,,,,averaged over threads,0,unbound,total.requestlist.avg
+unbound.total.requestlist.max,gauge,,,,the maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
```
```suggestion
unbound.total.requestlist.max,gauge,,,,The maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.num.recursivereplies,count,,,,summed over threads,1,unbound,total.num.recursivereplies
+unbound.total.requestlist.avg,gauge,,,,averaged over threads,0,unbound,total.requestlist.avg
+unbound.total.requestlist.max,gauge,,,,the maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
+unbound.total.requestlist.overwritten,gauge,,,,summed over threads,1,unbound,total.requestlist.overwritten
```
```suggestion
unbound.total.requestlist.overwritten,gauge,,request,,The total number of requests overwritten summed over threads,1,unbound,total.requestlist.overwritten
```
?

dbyron0 replied:
How's https://github.com/DataDog/integrations-extras/pull/451/commits/d0b247a56380a85da859f8894357f9596f5f7ddd ?

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.requestlist.avg,gauge,,,,averaged over threads,0,unbound,total.requestlist.avg
+unbound.total.requestlist.max,gauge,,,,the maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
+unbound.total.requestlist.overwritten,gauge,,,,summed over threads,1,unbound,total.requestlist.overwritten
+unbound.total.requestlist.exceeded,gauge,,,,summed over threads,1,unbound,total.requestlist.exceeded
```
```suggestion
unbound.total.requestlist.exceeded,gauge,,request,,The total number of requests exceeded summed over threads,1,unbound,total.requestlist.exceeded
```
?

dbyron0 replied:
Or https://github.com/DataDog/integrations-extras/pull/451/commits/cbf1a859611b918a74bbbc754b93ff1ad146b0c2 ?

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.requestlist.max,gauge,,,,the maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
+unbound.total.requestlist.overwritten,gauge,,,,summed over threads,1,unbound,total.requestlist.overwritten
+unbound.total.requestlist.exceeded,gauge,,,,summed over threads,1,unbound,total.requestlist.exceeded
+unbound.total.requestlist.current.all,gauge,,,,summed over threads,1,unbound,total.requestlist.current.all
```
Can you provide a better description?

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.requestlist.exceeded,gauge,,,,summed over threads,1,unbound,total.requestlist.exceeded
+unbound.total.requestlist.current.all,gauge,,,,summed over threads,1,unbound,total.requestlist.current.all
+unbound.total.recursion.time.avg,gauge,,,,averaged over threads,0,unbound,total.recursion.time.avg
+unbound.total.recursion.time.median,gauge,,,,averaged over threads,0,unbound,total.recursion.time.median
```
Can you provide a better description for the 3 lines above?

dbyron0 replied:
I'd love to find a good place to put this in the code or docs (or ideally right in metadata.csv, but I don't think that works), but all these descriptions (except a few from older versions that have since disappeared) came from https://nlnetlabs.nl/documentation/unbound/unbound-control/.

The full description for the per-thread recursion.time.median is:
```
              The  median  of  the  time it took to answer queries that needed
              recursive processing.  The median means that  50%  of  the  user
              queries  were  answered  in less than this time.  Because of big
              outliers (usually queries to non responsive servers), the  aver-
              age  can be bigger than the median.  This median has been calcu-
              lated by interpolation from a histogram.
```

Here's a crack at improving the descriptions: https://github.com/DataDog/integrations-extras/pull/451/commits/19a210c071ac7ddc3f9f595ddb9b560a61680780

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.requestlist.current.all,gauge,,,,summed over threads,1,unbound,total.requestlist.current.all
+unbound.total.recursion.time.avg,gauge,,,,averaged over threads,0,unbound,total.recursion.time.avg
+unbound.total.recursion.time.median,gauge,,,,averaged over threads,0,unbound,total.recursion.time.median
+unbound.time.now,gauge,,,,current time in seconds since 1970,1,unbound,time.now
```
```suggestion
unbound.time.now,gauge,,second,,The current time in seconds since 1970,1,unbound,time.now
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.recursion.time.avg,gauge,,,,averaged over threads,0,unbound,total.recursion.time.avg
+unbound.total.recursion.time.median,gauge,,,,averaged over threads,0,unbound,total.recursion.time.median
+unbound.time.now,gauge,,,,current time in seconds since 1970,1,unbound,time.now
+unbound.time.up,gauge,,,,uptime since server boot in seconds,0,unbound,time.up
```
```suggestion
unbound.time.up,gauge,,second,,The uptime since server boot in seconds,0,unbound,time.up
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.total.recursion.time.median,gauge,,,,averaged over threads,0,unbound,total.recursion.time.median
+unbound.time.now,gauge,,,,current time in seconds since 1970,1,unbound,time.now
+unbound.time.up,gauge,,,,uptime since server boot in seconds,0,unbound,time.up
+unbound.time.elapsed,gauge,,,,time since last statistics printout (in seconds),0,unbound,time.elapsed
```
```suggestion
unbound.time.elapsed,gauge,,second,,The time since last statistics printout in seconds,0,unbound,time.elapsed
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.time.now,gauge,,,,current time in seconds since 1970,1,unbound,time.now
+unbound.time.up,gauge,,,,uptime since server boot in seconds,0,unbound,time.up
+unbound.time.elapsed,gauge,,,,time since last statistics printout (in seconds),0,unbound,time.elapsed
+unbound.mem.total.sbrk,gauge,,,,If sbrk(2) is available an estimate of the heap size of the program in number of bytes,0,unbound,mem.total.sbrk
```
```suggestion
unbound.mem.total.sbrk,gauge,,byte,,If sbrk(2) is available an estimate of the heap size of the program in bytes,0,unbound,mem.total.sbrk
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.time.up,gauge,,,,uptime since server boot in seconds,0,unbound,time.up
+unbound.time.elapsed,gauge,,,,time since last statistics printout (in seconds),0,unbound,time.elapsed
+unbound.mem.total.sbrk,gauge,,,,If sbrk(2) is available an estimate of the heap size of the program in number of bytes,0,unbound,mem.total.sbrk
+unbound.mem.cache.rrset,gauge,,,,Memory in bytes in use by the RRset cache,0,unbound,mem.cache.rrset
```
```suggestion
unbound.mem.cache.rrset,gauge,,byte,,Memory in bytes in use by the RRset cache,0,unbound,mem.cache.rrset
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.time.elapsed,gauge,,,,time since last statistics printout (in seconds),0,unbound,time.elapsed
+unbound.mem.total.sbrk,gauge,,,,If sbrk(2) is available an estimate of the heap size of the program in number of bytes,0,unbound,mem.total.sbrk
+unbound.mem.cache.rrset,gauge,,,,Memory in bytes in use by the RRset cache,0,unbound,mem.cache.rrset
+unbound.mem.cache.message,gauge,,,,Memory in bytes in use by the message cache,0,unbound,mem.cache.message
```
```suggestion
unbound.mem.cache.message,gauge,,byte,,Memory in bytes in use by the message cache,0,unbound,mem.cache.message
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.mem.cache.rrset,gauge,,,,Memory in bytes in use by the RRset cache,0,unbound,mem.cache.rrset
+unbound.mem.cache.message,gauge,,,,Memory in bytes in use by the message cache,0,unbound,mem.cache.message
+unbound.mem.cache.dnscrypt_shared_secret,gauge,,,,Memory in bytes in use by the dnscrypt shared secrets cache,0,unbound,mem.cache.dnscrypt_shared_secret
+unbound.mem.cache.dnscrypt_nonce,gauge,,,,Memory in bytes in use by the dnscrypt nonce cache,0,unbound,mem.cache.dnscrypt_nonce
```
```suggestion
unbound.mem.cache.dnscrypt_nonce,gauge,,byte,,Memory in bytes in use by the dnscrypt nonce cache,0,unbound,mem.cache.dnscrypt_nonce
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.mem.cache.message,gauge,,,,Memory in bytes in use by the message cache,0,unbound,mem.cache.message
+unbound.mem.cache.dnscrypt_shared_secret,gauge,,,,Memory in bytes in use by the dnscrypt shared secrets cache,0,unbound,mem.cache.dnscrypt_shared_secret
+unbound.mem.cache.dnscrypt_nonce,gauge,,,,Memory in bytes in use by the dnscrypt nonce cache,0,unbound,mem.cache.dnscrypt_nonce
+unbound.mem.mod.iterator,gauge,,,,Memory in bytes in use by the iterator module,0,unbound,mem.mod.iterator
```
```suggestion
unbound.mem.mod.iterator,gauge,,byte,,Memory in bytes in use by the iterator module,0,unbound,mem.mod.iterator
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.mem.cache.dnscrypt_shared_secret,gauge,,,,Memory in bytes in use by the dnscrypt shared secrets cache,0,unbound,mem.cache.dnscrypt_shared_secret
+unbound.mem.cache.dnscrypt_nonce,gauge,,,,Memory in bytes in use by the dnscrypt nonce cache,0,unbound,mem.cache.dnscrypt_nonce
+unbound.mem.mod.iterator,gauge,,,,Memory in bytes in use by the iterator module,0,unbound,mem.mod.iterator
+unbound.mem.mod.validator,gauge,,,,Memory in bytes in use by the validator module. Includes the key cache and negative cache.,0,unbound,mem.mod.validator
```
```suggestion
unbound.mem.mod.validator,gauge,,byte,,Memory in bytes in use by the validator module. Includes the key cache and negative cache.,0,unbound,mem.mod.validator
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.mem.cache.dnscrypt_nonce,gauge,,,,Memory in bytes in use by the dnscrypt nonce cache,0,unbound,mem.cache.dnscrypt_nonce
+unbound.mem.mod.iterator,gauge,,,,Memory in bytes in use by the iterator module,0,unbound,mem.mod.iterator
+unbound.mem.mod.validator,gauge,,,,Memory in bytes in use by the validator module. Includes the key cache and negative cache.,0,unbound,mem.mod.validator
+unbound.mem.streamwait,gauge,,,,Memory in bytes in used by the TCP and TLS stream wait buffers. These are answers waiting to be written back to the clients.,0,unbound,mem.streamwait
```
```suggestion
unbound.mem.streamwait,gauge,,byte,,Memory in bytes in use by the TCP and TLS stream wait buffers. These are answers waiting to be written back to the clients.,0,unbound,mem.streamwait
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.num.answer.rcode,count,,,,The number of answers to queries from cache or from recursion by return code,1,unbound,num.answer.rcode
+unbound.num.answer.rcode.nodata,count,,,,The number of answers to queries that had the pseudo return code nodata,1,unbound,num.answer.rcode.nodata
+unbound.num.answer.secure,count,,,,The number of answers that were secure,1,unbound,num.answer.secure
+unbound.num.answer.bogus,count,,,,The number  of  answers  that were bogus,1,unbound,num.answer.bogus
```
```suggestion
unbound.num.answer.bogus,count,,,,The number of answers that were bogus,1,unbound,num.answer.bogus
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.num.rrset.bogus,count,,,,The number of rrsets marked bogus by the validator,1,unbound,num.rrset.bogus
+unbound.unwanted.queries,count,,query,,The number of queries that were refused or dropped because they failed the access control settings,1,unbound,unwanted.queries
+unbound.unwanted.replies,count,,,,The number of replies that were unwanted or unsolicited,1,unbound,unwanted.replies
+unbound.msg.cache.count,count,,,,The number of items (DNS replies) in the message cache,0,unbound,msg.cache.count
```
```suggestion
unbound.msg.cache.count,count,,item,,The number of items (DNS replies) in the message cache,0,unbound,msg.cache.count
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.unwanted.replies,count,,,,The number of replies that were unwanted or unsolicited,1,unbound,unwanted.replies
+unbound.msg.cache.count,count,,,,The number of items (DNS replies) in the message cache,0,unbound,msg.cache.count
+unbound.rrset.cache.count,count,,,,The number of RRsets in the rrset cache,0,unbound,rrset.cache.count
+unbound.infra.cache.count,count,,,,The number of items in the infra cache,0,unbound,infra.cache.count
```
```suggestion
unbound.infra.cache.count,count,,item,,The number of items in the infra cache,0,unbound,infra.cache.count
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.msg.cache.count,count,,,,The number of items (DNS replies) in the message cache,0,unbound,msg.cache.count
+unbound.rrset.cache.count,count,,,,The number of RRsets in the rrset cache,0,unbound,rrset.cache.count
+unbound.infra.cache.count,count,,,,The number of items in the infra cache,0,unbound,infra.cache.count
+unbound.key.cache.count,count,,,,The number of items in the key cache,0,unbound,key.cache.count
```
```suggestion
unbound.key.cache.count,count,,item,,The number of items in the key cache,0,unbound,key.cache.count
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.rrset.cache.count,count,,,,The number of RRsets in the rrset cache,0,unbound,rrset.cache.count
+unbound.infra.cache.count,count,,,,The number of items in the infra cache,0,unbound,infra.cache.count
+unbound.key.cache.count,count,,,,The number of items in the key cache,0,unbound,key.cache.count
+unbound.dnscrypt_shared_secret.cache.count,count,,,,The number of items in the shared secret cache,0,unbound,dnscrypt_shared_secret.cache.count
```
```suggestion
unbound.dnscrypt_shared_secret.cache.count,count,,item,,The number of items in the shared secret cache,0,unbound,dnscrypt_shared_secret.cache.count
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.infra.cache.count,count,,,,The number of items in the infra cache,0,unbound,infra.cache.count
+unbound.key.cache.count,count,,,,The number of items in the key cache,0,unbound,key.cache.count
+unbound.dnscrypt_shared_secret.cache.count,count,,,,The number of items in the shared secret cache,0,unbound,dnscrypt_shared_secret.cache.count
+unbound.dnscrypt_nonce.cache.count,count,,,,The number of items in the client nonce cache,0,unbound,dnscrypt_nonce.cache.count
```
```suggestion
unbound.dnscrypt_nonce.cache.count,count,,item,,The number of items in the client nonce cache,0,unbound,dnscrypt_nonce.cache.count
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.num.query.authzone.down,count,,query,,The number of queries for downstream answered from auth-zone data,1,unbound,num.query.authzone.down
+unbound.num.query.aggressive.NOERROR,count,,query,,The number of queries answered using cached NSEC records with NODATA RCODE,1,unbound,num.query.aggressive.NOERROR
+unbound.num.query.aggressive.NXDOMAIN,count,,query,,The number of queries answered using cached NSEC records with NXDOMAIN RCODE,1,unbound,num.query.aggressive.NXDOMAIN
+unbound.num.query.subnet,count,,query,,The number of queries that got an answer that contained EDNS client subnet data,1,unbound,num.query.subnet
```
```suggestion
unbound.num.query.subnet,count,,query,,The number of queries which received an answer and contained EDNS client subnet data,1,unbound,num.query.subnet
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.num.query.aggressive.NOERROR,count,,query,,The number of queries answered using cached NSEC records with NODATA RCODE,1,unbound,num.query.aggressive.NOERROR
+unbound.num.query.aggressive.NXDOMAIN,count,,query,,The number of queries answered using cached NSEC records with NXDOMAIN RCODE,1,unbound,num.query.aggressive.NXDOMAIN
+unbound.num.query.subnet,count,,query,,The number of queries that got an answer that contained EDNS client subnet data,1,unbound,num.query.subnet
+unbound.num.query.subnet_cache,count,,query,,The number of queries answered from the edns client subnet cache,1,unbound,num.query.subnet_cache
```
```suggestion
unbound.num.query.subnet_cache,count,,query,,The number of queries answered from the EDNS client subnet cache,1,unbound,num.query.subnet_cache
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.requestlist.max,gauge,,request,,Maximum size attained by the internal recursive processing request list,1,unbound,thread.requestlist.max
+unbound.thread.requestlist.overwritten,count,,,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
+unbound.thread.requestlist.exceeded,gauge,,query,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
+unbound.thread.requestlist.current.all,gauge,,,,Current size of the request list,0,unbound,thread.requestlist.current
```
```suggestion
unbound.thread.requestlist.current.all,gauge,,request,,Current size of the request list,0,unbound,thread.requestlist.current
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,82 @@
+unbound.thread.requestlist.overwritten,count,,,,The number of requests in the request list that were overwritten by newer entries,1,unbound,thread.requestlist.overwritten
+unbound.thread.requestlist.exceeded,gauge,,query,, Queries that were dropped because the request list was full,1,unbound,thread.requestlist.exceeded
+unbound.thread.requestlist.current.all,gauge,,,,Current size of the request list,0,unbound,thread.requestlist.current
+unbound.thread.requestlist.current.user,gauge,,,,Current size of the request list (only the requests from client queries),0,unbound,thread.requestlist.current.user
```
```suggestion
unbound.thread.requestlist.current.user,gauge,,request,,Current size of the request list (only the requests from client queries),0,unbound,thread.requestlist.current.user
```

dbyron0 commented:
OK, I'm punting on an integration test.  Too much struggle getting a properly-configured unbound to come up in a container, and then too many code changes only for the sake of the test run do something like
```
docker-compose exec -T unbound unbound-control stats
```

So I manually collected some stats from 1.9.2 and turned those into fixtures / additional tests...even found a bug along the way :)  https://github.com/DataDog/integrations-extras/pull/451/commits/b0a9842b00a950530e1b8fedddba50969bb5bbb2

At this point I don't have any more changes queued up.  Thanks for the review so far.  Looking forward to more.

dbyron0 commented:
That said, I haven't exercised this with a real agent yet....

dabcoder commented:
> That said, I haven't exercised this with a real agent yet....

If you have a chance to test this with the latest version of Agent 6, that would be great indeed.

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,83 @@
+unbound.total.num.recursivereplies,count,,,,The total number of recursive replies summed over threads,1,unbound,total.num.recursivereplies
+unbound.total.requestlist.avg,gauge,,request,,The total number of requests averaged over threads,0,unbound,total.requestlist.avg
+unbound.total.requestlist.max,gauge,,,,The maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
+unbound.total.requestlist.overwritten,gauge,,,,The number of requests in the request list that were overwritten by newer entries summed over threads,1,unbound,total.requestlist.overwritten
```
```suggestion
unbound.total.requestlist.overwritten,gauge,,request,,The number of requests in the request list that were overwritten by newer entries summed over threads,1,unbound,total.requestlist.overwritten
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,83 @@
+unbound.total.requestlist.avg,gauge,,request,,The total number of requests averaged over threads,0,unbound,total.requestlist.avg
+unbound.total.requestlist.max,gauge,,,,The maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
+unbound.total.requestlist.overwritten,gauge,,,,The number of requests in the request list that were overwritten by newer entries summed over threads,1,unbound,total.requestlist.overwritten
+unbound.total.requestlist.exceeded,gauge,,,,The total number of queries that were dropped because the request list was full summed over threads,1,unbound,total.requestlist.exceeded
```
```suggestion
unbound.total.requestlist.exceeded,gauge,,query,,The total number of queries that were dropped because the request list was full summed over threads,1,unbound,total.requestlist.exceeded
```

ruthnaebeck commented on unbound/metadata.csv:
```diff
@@ -0,0 +1,83 @@
+unbound.total.requestlist.max,gauge,,,,The maximum of the thread requestlist.max values,1,unbound,total.requestlist.max
+unbound.total.requestlist.overwritten,gauge,,,,The number of requests in the request list that were overwritten by newer entries summed over threads,1,unbound,total.requestlist.overwritten
+unbound.total.requestlist.exceeded,gauge,,,,The total number of queries that were dropped because the request list was full summed over threads,1,unbound,total.requestlist.exceeded
+unbound.total.requestlist.current.all,gauge,,,,The current size of all threads' request lists,1,unbound,total.requestlist.current.all
```
If size = number of requests, add request as unit name
If size is something like bytes, add that as the unit name

dbyron0 replied:
Here's a shot at populating more units: https://github.com/DataDog/integrations-extras/pull/451/commits/63950178628d3de6118d7f643433097eef6e9905

dbyron0 commented:
Finally getting back to this, and struggling to get the check running with a real agent.
I'm using datadog agent 6.12.1 and following the instructions at https://docs.datadoghq.com/developers/integrations/new_check_howto/ except I'm cloning my fork and checking out this branch...

It all seems so promising, but then don't see any evidence that the check is actually getting installed.

```
# sudo -u dd-agent datadog-agent integration install -v -w /opt/datadog-agent/integrations-extras/unbound/dist/datadog_unbound-0.0.1-py2.py3-none-any.whl
For your security, only use this to install wheels containing an Agent integration and coming from a known source. The Agent cannot perform any verification on local wheels.
Ignoring indexes: https://pypi.org/simple
Created temporary directory: /tmp/pip-ephem-wheel-cache-UDDp1L
Created temporary directory: /tmp/pip-req-tracker-jkSXvw
Created requirements tracker '/tmp/pip-req-tracker-jkSXvw'
Created temporary directory: /tmp/pip-install-X93St_
Processing ./integrations-extras/unbound/dist/datadog_unbound-0.0.1-py2.py3-none-any.whl
  Added datadog-unbound==0.0.1 from file:///opt/datadog-agent/integrations-extras/unbound/dist/datadog_unbound-0.0.1-py2.py3-none-any.whl to build tracker '/tmp/pip-req-tracker-jkSXvw'
  Removed datadog-unbound==0.0.1 from file:///opt/datadog-agent/integrations-extras/unbound/dist/datadog_unbound-0.0.1-py2.py3-none-any.whl from build tracker '/tmp/pip-req-tracker-jkSXvw'
Installing collected packages: datadog-unbound

Successfully installed datadog-unbound-0.0.1
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-jkSXvw'
Successfully copied configuration file conf.yaml.example
Successfully completed the installation of datadog-unbound
```
and then I start the agent:
```
# systemctl start datadog-agent
```
but then, no sign of the unbound collector.  Am I right to expect to see it here?
```
# datadog-agent status
Getting the status from the agent.

===============
Agent (v6.12.1)
===============

  Status date: 2019-07-05 22:52:36.197620 UTC
  Agent start: 2019-07-05 22:51:50.662276 UTC
  Pid: 3034
  Python Version: 2.7.16
  Check Runners: 4
  Log Level: warn

  Paths
  =====
    Config File: /etc/datadog-agent/datadog.yaml
    conf.d: /etc/datadog-agent/conf.d
    checks.d: /etc/datadog-agent/checks.d

  Clocks
  ======
    NTP offset: 96µs
    System UTC time: 2019-07-05 22:52:36.197620 UTC

  Host Info
  =========
    bootTime: 2019-07-05 22:50:39.000000 UTC
    kernelVersion: 4.9.0-9-amd64
    os: linux
    platform: debian
    platformFamily: debian
    platformVersion: 9.9
    procs: 140
    uptime: 1m13s
    virtualizationRole: guest
    virtualizationSystem: xen

  Hostnames
  =========
    ec2-hostname: XXX.us-west-2.compute.internal
    hostname: YYY
    instance-id: YYY
    socket-fqdn: XXX.us-west-2.compute.internal.
    socket-hostname: XXX
    hostname provider: aws
    unused hostname providers:
      configuration/environment: hostname is empty
      gce: unable to retrieve hostname from GCE: status code 404 trying to GET http://169.254.169.254/computeMetadata/v1/instance/hostname

=========
Collector
=========



  Running Checks
  ==============
    
    cpu
    ---
      Instance ID: cpu [OK]
      Total Runs: 3
      Metric Samples: Last Run: 6, Total: 12
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    disk (2.2.0)
    ------------
      Instance ID: disk:e5dffb8bef24336f [OK]
      Total Runs: 3
      Metric Samples: Last Run: 60, Total: 180
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 49ms
      
    
    file_handle
    -----------
      Instance ID: file_handle [OK]
      Total Runs: 2
      Metric Samples: Last Run: 5, Total: 10
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    io
    --
      Instance ID: io [OK]
      Total Runs: 3
      Metric Samples: Last Run: 26, Total: 60
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    load
    ----
      Instance ID: load [OK]
      Total Runs: 2
      Metric Samples: Last Run: 6, Total: 12
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    memory
    ------
      Instance ID: memory [OK]
      Total Runs: 3
      Metric Samples: Last Run: 17, Total: 51
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    network (1.11.0)
    ----------------
      Instance ID: network:e0204ad63d43c949 [OK]
      Total Runs: 3
      Metric Samples: Last Run: 26, Total: 78
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
    
    ntp
    ---
      Instance ID: ntp:b4579e02d1981c12 [OK]
      Total Runs: 3
      Metric Samples: Last Run: 1, Total: 3
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 3
      Average Execution Time : 87ms
      
    
    uptime
    ------
      Instance ID: uptime [OK]
      Total Runs: 3
      Metric Samples: Last Run: 1, Total: 3
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      
========
JMXFetch
========

  Initialized checks
  ==================
    no checks
    
  Failed checks
  =============
    no checks
    
=========
Forwarder
=========

  Transactions
  ============
    CheckRunsV1: 2
    Dropped: 0
    DroppedOnInput: 0
    Events: 0
    HostMetadata: 0
    IntakeV1: 2
    Metadata: 0
    Requeued: 0
    Retried: 0
    RetryQueueSize: 0
    Series: 0
    ServiceChecks: 0
    SketchSeries: 0
    Success: 6
    TimeseriesV1: 2

  API Keys status
  ===============
    API key ending with ece11: API Key valid

==========
Endpoints
==========
  https://app.datadoghq.com - API Key ending with:
      - ece11

==========
Logs Agent
==========

  Logs Agent is not running

=========
Aggregator
=========
  Checks Metric Sample: 454
  Dogstatsd Metric Sample: 1
  Event: 1
  Events Flushed: 1
  Number Of Flushes: 2
  Series Flushed: 171
  Service Check: 26
  Service Checks Flushed: 20

=========
DogStatsD
=========
  Event Packets: 0
  Event Parse Errors: 0
  Metric Packets: 0
  Metric Parse Errors: 0
  Service Check Packets: 0
  Service Check Parse Errors: 0
  Udp Packet Reading Errors: 0
  Udp Packets: 1
  Uds Origin Detection Errors: 0
  Uds Packet Reading Errors: 0
  Uds Packets: 0
```
I also don't see anything at https://app.datadoghq.com/check/summary, nor anything in /var/log/datadog/agent.log.  No other files in /var/log/datadog.

Thanks for your help.

dbyron0 commented:
🤦‍♂ was missing the step to copy /etc/datadog-agent/conf.d/unbound.d/conf.yaml.example to conf.yaml.  Now that that's there, the check is showing up in `datadog-agent status` and datadog.agent.check_status shows up as critical for check:unbound at https://app.datadoghq.com/check/summary since I haven't actually installed unbound yet...time to do that.

dbyron0 commented:
For the record, `datadog-agent status` shows this for unbound:
```
    unbound (0.0.1)
    ---------------
      Instance ID: unbound:d884b5186b651429 [ERROR]
      Total Runs: 2
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Error: 'unbound-control' executable not found
      Traceback (most recent call last):
        File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/datadog_checks/base/checks/base.py", line 884, in run
          self.check(instance)
        File "/opt/datadog-agent/embedded/lib/python2.7/site-packages/datadog_checks/unbound/unbound.py", line 28, in check
          raise ConfigurationError("'" + unbound_control + "' executable not found")
      ConfigurationError: 'unbound-control' executable not found
```

dbyron0 commented:
OK, I've verified that this works with datadog agent version 6.12.1 and unbound version 1.6.0 on debian stretch.

dbyron0 commented:
Would love a (final?) review.  I'm planning to do a bunch of commit squashing.  Any preference for whether that happens before or after the review?

hithwen commented:
Hi @dbyron0 just confirming. You are not seeing the `'unbound-control' executable not found` error anymore when you run `datadog-agent status`?

dbyron0 commented:
When unbound is actually installed, no.

dbyron0 commented on unbound/datadog_checks/unbound/data/conf.yaml.example:
```diff
@@ -0,0 +1,43 @@
+    # use_sudo: true
+
+    ## @param stats_command - string - optional
+    ## unbound-control command to execute to gather stats, defaults to stats
```
I'm pondering changing the default here to stats_noreset.  See https://nlnetlabs.nl/documentation/unbound/unbound-control/ for a description of stats and stats_noreset.  Trying to think of when we'd ever want to reset the counters.  Thoughts?

dabcoder replied:
> Trying to think of when we'd ever want to reset the counters. Thoughts?
Your call I think... I am not familiar with what the internal counters do, do you have an example in mind we could look at? To be fair you can leave it to `stats` by default. Can be reconsidered later one if need be.

dbyron0 replied:
unbound's counters are the source of the metrics reported here.  https://nlnetlabs.nl/documentation/unbound/howto-statistics/ describes setting `statistics-cumulative` to `no` which is what makes it OK to use the stats command here.

So, what's the easier thing?  To leave stats here, and require people to set `statistics-cumulative`, or use stats_noreset here and then maybe it doesn't matter what `statistics-cumulative` is set to.  I can't think of a reason why resetting the counters is helpful, except that eventually I suppose they'll wrap.

dabcoder replied:
Let's go with `stats` by default. No change needed then.

dbyron0 replied:
Sure.

ofek commented on unbound/datadog_checks/unbound/unbound.py:
```diff
@@ -0,0 +1,283 @@
+import os
+import re
+
+from datadog_checks.base import AgentCheck, ConfigurationError
```
```suggestion
from datadog_checks.base import AgentCheck, ConfigurationError, is_affirmative
```

ofek commented on unbound/datadog_checks/unbound/unbound.py:
```diff
@@ -0,0 +1,283 @@
+import re
+
+from datadog_checks.base import AgentCheck, ConfigurationError
+from datadog_checks.config import _is_affirmative
```
```suggestion
```

ofek commented on unbound/datadog_checks/unbound/unbound.py:
```diff
@@ -0,0 +1,283 @@
+
+    def check(self, instance):
+
+        use_sudo = _is_affirmative(instance.get('use_sudo', False))
```
```suggestion
        use_sudo = is_affirmative(instance.get('use_sudo', False))
```

ofek commented on unbound/datadog_checks/unbound/unbound.py:
```diff
@@ -0,0 +1,283 @@
+            command.append('sudo')
+
+        if not which(unbound_control, use_sudo, self.log):
+            raise ConfigurationError("'" + unbound_control + "' executable not found")
```
```suggestion
            raise ConfigurationError('executable not found: {}'.format(unbound_control))
```

ofek commented on unbound/datadog_checks/unbound/unbound.py:
```diff
@@ -0,0 +1,283 @@
+
+        if not data:
+            self.service_check(
+                self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, message="unsable to parse stats", tags=tags
```
```suggestion
                self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, message="unable to parse stats", tags=tags
```

dbyron0 commented:
Great.  Seems like it's time to do some serious squashing...Here goes...
